state/relation.go

```diff
@@ -780,3 +781,54 @@ func (change relationSettingsCleanupChange) Prepare(db Database) ([]txn.Op, erro
 	return ops, nil
 
 }
+
+func relationApplicationSettingsKey(id int, ep Endpoint) string {
+	return fmt.Sprintf("%s#%s#%s", relationGlobalScope(id), ep.Role, ep.ApplicationName)
+}
+
+// ApplicationSettings returns the application-level settings for the
+// specified application in this relation.
+func (r *Relation) ApplicationSettings(app *Application) (map[string]interface{}, error) {
+	ep, err := r.Endpoint(app.Name())
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	applicationKey := relationApplicationSettingsKey(r.Id(), ep)
+	s, err := readSettings(r.st.db(), settingsC, applicationKey)
+	if err != nil {
+		return nil, errors.Annotatef(err, "relation %q application %q", r.String(), app.Name())
+	}
+	return s.Map(), nil
+}
+
+// UpdateApplicationSettings updates the given application's settings
+// in this relation. It requires a current leadership token.
+func (r *Relation) UpdateApplicationSettings(app *Application, token leadership.Token, updates map[string]interface{}) error {
+	// We can calculate the actual update ahead of time; it's not dependent
+	// upon the current state of the document. (*Writing* it should depend
+	// on document state, but that's handled below.)
+	ep, err := r.Endpoint(app.Name())
+	if err != nil {
+		return errors.Trace(err)
+	}
+	key := relationApplicationSettingsKey(r.Id(), ep)
+	err = updateLeaderSettings(r.st.db(), token, key, updates)
```

Comment:
the other place where we call updateLeadersettings in this PR we first take a copy of the updates

Reply:
That's because it has a `map[string]string` but `updateLeaderSettings` takes a `map[string]interface{}` (for consistency with relation unit settings). It's not copying to avoid mutation, just type conversion.